File: codes/julia/custom_error_and_error_handling.jl
println("\n# Custom Error and Error Handling in Julia")

function give_me_respect(argument::Any)::Any
    # Custom Error
    if (argument !== "respect") throw(ErrorException("Exception: You should give me \"respect\"!")) end
    return "Thank you for giving me \"respect\"!"
end

(() -> begin
    # Error Handling
    try
        local response::Any = give_me_respect("boo!")
        println(response)
    catch (error_exception)
        println(error_exception.msg)
    end
    println("I'm so sorry!")
end)()

(() -> begin
    # Without Error Handling
    local response::Any = give_me_respect("boo!")
    # this will never executed
    println(response)
    println("I'm so sorry!")
end)()

File: codes/r/custom_error_and_error_handling.r
cat("\n# Custom Error and Error Handling in R\n")

give_me_respect <- function(argument) {
    # Custom Error
    if (argument != 0) stop("Error: You should give me \"respect\"!")
    return("Thank you for giving me \"respect\"!")
}

# Error Handling
tryCatch(
    expr = {
        response <- give_me_respect("boo!")
        cat(paste(sep = "", response, "\n"))
    },
    error = function(an_error) {
        cat(paste(sep = "", conditionMessage(an_error), "\n"))
    }
)
cat(paste(sep = "", "I'm sorry!", "\n"))

# Without Error Handling
response <- give_me_respect("boo!")
# this will never executed
cat(paste(sep = "", response, "\n"))
cat(paste(sep = "", "I'm sorry!", "\n"))
